### PR TITLE
add infinite loop detection

### PR DESCRIPTION
**Pull request by JosiasAurel**
the sprig editor will throw an error if
1. there are more than 2000 iterations in a loop AND
2. the loop takes > 1.5s to complete

image

addresses #1091 

add environment variables to Vercel deployment
- [x] `PUBLIC_MAX_ITERATIONS`
- [x] `PUBLIC_MAX_LOOP_TIME_MS`

### CODE REVIEW

**Review by grymmy — CHANGES_REQUESTED** [review]
Please answer a few questions for me:

1. Is this performant? (will it slow down the execution of games?)  How do you know?
2. What is the UX when an infinite loop is detected?
3. Is this more likely to flag as inf loop when your computer is under load? 
4. Where do we intend on documenting this watchdog behavior such that people are not surprised when encountering this?

Separately (I won't demand this), it would be nice to have a few unit tests here proving it works.

File: src/lib/transform-detect-infinite-loop.ts [hunk 1]
```diff
@@ -0,0 +1,89 @@
+// Based on https://repl.it/site/blog/infinite-loops by Amjad Masad.
+
+export default function TransformDetectInfiniteLoop({ types: t }) {
+	const MAX_ITERATIONS = 2000;
```
Comment (grymmy) [inline src/lib/transform-detect-infinite-loop.ts:4]: Suggest moving these to environment variables (or to another constants definition that's higher-level in the project).

**Review by grymmy — COMMENTED** [review]
Can you also ensure that `debugger` statements continue to work w/ this?

**Comment by JosiasAurel** [comment]
> Please answer a few questions for me:
> 
> 1. Is this performant? (will it slow down the execution of games?)  How do you know?
> 2. What is the UX when an infinite loop is detected?
> 3. Is this more likely to flag as inf loop when your computer is under load?
> 4. Where do we intend on documenting this watchdog behavior such that people are not surprised when encountering this?
> 
> Separately (I won't demand this), it would be nice to have a few unit tests here proving it works.

1. I have tested these changes with a few games in the gallery and there was no noticeable slow down. 
2. This will show up as an error in the editor (see original comment)
3. I'm not sure
4. This will be added directly in the toolkit page in the editor 

image

**Comment by JosiasAurel** [comment]
> Can you also ensure that `debugger` statements continue to work w/ this?

debugger statements still work

**Comment by Goldenpig12345** [comment]
Oh, My Goodness! it's so nice to see you again! How's Anne doing? I
have seen her since she was three!
My doctor has recently said you haven't eaten enough turnips lately. Do you
know if this is true? I really miss you and we need to catch up soon. do
you still have my number? oh, it's okay if you don't I'll just send it to
you. my number is (802)-658-5667. oh ad I heard that your mother was
arrested could you tell me what happened? something about an adult diaper
chase at Walmart? anyway please get back to me soon and maybe we could do
lunch. Love You Lots.

File: docs/docs.md [hunk 2]
```diff
@@ -8,6 +8,23 @@ and sharing them with others. You can watch [this video](https://www.youtube.com
 
 Run games by hitting the `Run` button or pressing `Shift+Enter`.
 
+### Infinite loop detection
```
Comment (grymmy) [inline docs/docs.md:11]: Suggest cutting this.  Users will experience it intuitively and this probably shouldn't be above the "Getting Help" section.

**Review by grymmy — APPROVED** [review]

**Pull request merged**